docs/enhancements/unified-configuration-of-hive-controllers.md

```diff
@@ -0,0 +1,239 @@
+# Unified Configuration of Hive controllers
+
+## Summary
+
+Each Hive controller loads its configuration from a ConfigDir. The ConfigDir provides
+the controller a YAML serialized HiveConfig and all needed ConfigMaps and Secrets.
+Hive operator ensures the ConfigDir is setup with latest copy of HiveConfig and
+mounts all ConfigMaps and Secrets referenced in HiveConfig to ConfigDir so that the
+controller can read the contents from disk.
+
+## Motivation
+
+Currently each Hive controller decides how it accepts configuration like using Env vars,
+files on disk, k8s objects like Secrets fetched at runtime, or combinations of these.
+But the source of all these configuration is one object HiveConfig, and the Hive operator
+has the code to transform the HiveConfig contents to different destinations as expected
+by these Hive controllers. There is a lot of new code required in Hive operator and
+controller every time a new controller is added or new configuration is required. Also
+there is no standard way to debug the configuration because of this heterogeneity.
+
+To make sure it is easier to add new configuration, we must standardize the configuration
+to one way for all controllers.
+
+### Goals
+
+1. No additional code required to support new configuration that only depends on HiveConfig
+    content.
+2. Minimal code required to support new configuration that depends on data from
+    additional k8s objects like Secrets or ConfigMaps.
+3. Reduce different types of configuration sources.
+4. Remove decision making of how to configure the controllers.
+
+### Non-Goals
+
+## Proposal
+
+All hive controllers accept a directory (ConfigDir) for configuration. This directory
+provides,
+
+1. YAML serialized HiveConfig,
+2. Secrets specified in HiveConfig, and
+3. ConfigMaps specified in HiveConfig.
+
+```
+$CONFIG_DIR/
+    hiveconfig/
+        config.yaml
+    secrets/
+        <secret-1>/
+            <secret-1-key1>
+            <secret-1-key2>
+        <secret-2>/
+            <secret-2-key1>
+        ...
+    configmaps/
+        <configmap-1>/
+            <configmap-1-key1>
+        ...
```

Comment:
Secrets and ConfigMaps I'm not entirely sure what we gain here, as opposed to just reading what we need on the fly. Restarting the pod can be a pretty expensive operation in a heavily loaded Hive, and reading on the fly makes the local dev case quite a bit simpler, you just need the main hiveconfig file. 

 What benefits do you see to the complexity of setting up the tree of mounts, vs reading the ref of what we need from HiveConfig file on disk and loading that object with a kube client as needed?